Add predicates for subsets, supersets, and disjoint sets

thass0 opened the pull request.

thass0 committed:
Add `is_{proper_}subset` and `is_{proper_}superset`

thass0 committed:
Add `is_disjoint` method

Chriscbr commented on src/lib.rs:
```diff
@@ -517,6 +517,91 @@ impl<T: PrimInt + Unsigned> SparseSet<T> {
+    ///
+    /// This is the same as [`Self::is_proper_superset`] with the arguments flipped.
+    pub fn is_proper_subset(&self, other: &Self) -> bool {
+        self.is_subset(other) && self.inner.len() < other.inner.len()
```
Would it be faster to perform the size check first due to boolean short circuiting?

(in the case the result is `true`, both sides have to be computed so the order doesn't matter -- but in the case it's `false`, you'd rather perform the cheaper computation first)
```suggestion
        self.inner.len() < other.inner.len() && self.is_subset(other)
```

thass0 replied:
Yes, it would! I'll update both cases

Chriscbr commented on src/lib.rs:
```diff
@@ -517,6 +517,91 @@ impl<T: PrimInt + Unsigned> SparseSet<T> {
+    ///
+    /// This is the same as [`Self::is_proper_subset`] with the arguments flipped.
+    pub fn is_proper_superset(&self, other: &Self) -> bool {
+        self.is_superset(other) && self.inner.len() > other.inner.len()
```
See comment above

Chriscbr commented on src/lib.rs:
```diff
@@ -517,6 +517,91 @@ impl<T: PrimInt + Unsigned> SparseSet<T> {
+            }
+        }
+
+        for x in other.iter() {
+            if self.contains(x) {
+                return false;
+            }
+        }
+
+        true
```
I think only the first loop is needed, unless I'm mistaken
```suggestion
        true
```

thass0 replied:
Yes, you are correct, I think. If  the sets *A* and *B* are not disjoint, then they have some elements in common. Any element that they have in common can be found in *A* **and** in *B*, so, to run the `is_disjoint` check, it's sufficient to iterate over the elements of one set only. Regardless of which one, the common elements would be picked up.

thass0 committed:
Improve predicate speed

* Lengths checks offer a fast path in some cases
* Checking set membership both ways in `is_disjoint`
  is not necessary. To find out if there are common
  elements between two sets, it's enough to iterate
  over only one of the sets. If there are elements
  that they both contain, it doesn't matter which
  of the sets is used as a starting point.
